ocaml/lambda/translcore.ml

```diff
@@ -51,8 +51,16 @@ let layout_must_be_value loc layout =
 
 (* CR layouts v2: In the places where this is used, we want to allow any (the
    left of a semicolon and loop bodies).  So we want this instead of the usual
-   sanity check for value.  *)
+   sanity check for value.  But we still default to value before checking for
+   void, to allow for sort variables arising in situations like
+
+     let foo () = raise Foo; ()
+
+   When this sanity check is removed, consider whether we are still defaulting
+   appropriately.
+*)
 let layout_must_not_be_void loc layout =
+  Layout.default_to_value layout;
```

Comment:
This is a small perf concern, but could be fixed by refactoring `default_to_value`.

Reply:
Agreed.

I just pushed a commit that refactors the `repr` stuff in layouts, implementing the observation we made a while ago that if we have a separate version of `repr` for the case where we're defaulting, it needn't allocate.  While I was there, I made a couple other simplifications:

- We only ever default to value.  Now you don't have a choice, which simplifies things a little.  While we might change the default to `void` later, we have no plans for a version that defaults to different things in different spots.
- Once those simplifications were made, there was no difference between `repr` and `get`.  So now we just have `get` and `get_default_layout`.  I shuffled the functions around a little to put the related "elimination" and "defaulting" functions together.

I think the result is a nice overall simplification, but yell at me if you think otherwise.

Reply:
Happy here. Thanks for doing this.